src/rust/engine/concrete_time/src/lib.rs

```diff
@@ -0,0 +1,187 @@
+#![allow(clippy::mutex_atomic)]
+
+use log::warn;
+use serde_derive::Serialize;
+
+/// A concrete data representation of a duration.
+/// Unlike std::time::Duration, it doesn't hide how the time is stored as the purpose of this
+/// `struct` is to expose it.
+///
+/// This type can be serialized with serde.
+///
+/// This type can be converted from and into a `std::time::Duration` as this should be the goto
+/// data representation for a `Duration` when one isn't concerned about serialization.
+///
+/// It can be used to represent a timestamp (as a duration since the unix epoch) or simply a
+/// duration between two arbitrary timestamps.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
+pub struct Duration {
+  /// How many seconds did this `Duration` last?
+  pub secs: u64,
+  /// How many sub-second nanoseconds did this `Duration` last?
+  pub nanos: u32,
+}
+
+impl Duration {
+  /// Construct a new duration with `secs` seconds and `nanos` nanoseconds
+  pub fn new(secs: u64, nanos: u32) -> Self {
+    Self { secs, nanos }
+  }
+}
+
+impl From<std::time::Duration> for Duration {
+  fn from(duration: std::time::Duration) -> Self {
+    Self {
+      secs: duration.as_secs(),
+      nanos: duration.subsec_nanos(),
+    }
+  }
+}
+
+impl Into<std::time::Duration> for Duration {
+  fn into(self) -> std::time::Duration {
+    std::time::Duration::new(self.secs, self.nanos)
+  }
+}
+
+/// A timespan
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize)]
+pub struct TimeSpan {
+  /// Duration since the UNIX_EPOCH
+  pub start: Duration,
+  /// Duration since `start`
+  pub duration: Duration,
+}
+
+impl TimeSpan {
+  fn since_epoch(time: &std::time::SystemTime) -> std::time::Duration {
+    time
+      .duration_since(std::time::UNIX_EPOCH)
+      .expect("Surely you're not before the unix epoch?")
+  }
+
+  /// Construct a TimeSpan that started at `start` and ends now.
+  pub fn since(start: &std::time::SystemTime) -> TimeSpan {
+    let start = Self::since_epoch(start);
+    let duration = Self::since_epoch(&std::time::SystemTime::now()) - start;
+    TimeSpan {
+      start: start.into(),
+      duration: duration.into(),
+    }
+  }
+
+  fn std_duration_from_protobuf_timestamp(
+    t: &protobuf::well_known_types::Timestamp,
+  ) -> std::time::Duration {
+    std::time::Duration::new(t.seconds as u64, t.nanos as u32)
+  }
+
+  /// Construct a `TimeSpan` given a start and an end `Timestamp` from protobuf
+  pub fn from_start_and_end(
+    start: &protobuf::well_known_types::Timestamp,
+    end: &protobuf::well_known_types::Timestamp,
+    time_span_description: &str,
+  ) -> Option<Self> {
+    let start = Self::std_duration_from_protobuf_timestamp(start);
+    let end = Self::std_duration_from_protobuf_timestamp(end);
+    let time_span = end.checked_sub(start).map(|duration| TimeSpan {
+      start: start.into(),
+      duration: duration.into(),
+    });
+    if time_span.is_none() {
+      warn!(
```

Comment:
Should this just crash? I'll look through the usages, but it doesn't look like returning `None` is what we want to do.

I think I'd prefer just crashing, or returning a `Result<Timespan, String>` or something like this.

Would also be okay to leave it as a TODO:

```suggestion
     // TODO Make this return a Result<> or panic instead of just warning.
      warn!(
```

Reply:
I see your point, and it may be worth doing in a separate PR. The reason I'm only warning here is because this change is a simple refactoring, so I'm not aiming to change the behaviour too significantly. You can find the old behaviour in src/rust/engine/process_execution/src/remote.rs (duplicated a few times).

Reply:
I could return an Err from here and warn at the call site, though.

Reply:
Addressed in an extra commit